### ISSUE TITLE
ts__default.createIdentifier is not a function

### ISSUE THREAD

**Issue by m-mcgowan** [nonara/ts-patch#149]
Hi,

I'm fairly experienced with typescript, but less so with the transforms and the tooling landscape there. 

I'd like to use [`ts-type-checked`](https://www.npmjs.com/package/ts-type-checked) to automate creation of type guards for validating external input.  That transform was written for `ttypescript`, but hearing that `ts-patch` is a drop-in replacement, I gave that a try, as it supports TS 5.

When I run `tspc` on my project, I get this error:

```
TypeError: ts__default.createIdentifier is not a function
    at /Users/mat/e/firebase/backend/functions/node_modules/ts-type-checked/transformer.js:958:57
    at tspWrappedFactory (evalmachine.<anonymous>:223:31)
    at transformSourceFileOrBundle (evalmachine.<anonymous>:90663:51)
    at transformation (evalmachine.<anonymous>:114095:16)
    at transformRoot (evalmachine.<anonymous>:114118:73)
    at transformNodes (evalmachine.<anonymous>:114103:72)
    at emitJsFileOrBundle (evalmachine.<anonymous>:114676:26)
    at emitSourceFileOrBundle (evalmachine.<anonymous>:114631:7)
    at forEachEmittedFile (evalmachine.<anonymous>:114379:26)
    at emitFiles (evalmachine.<anonymous>:114605:5)
 ```

The `ts__default` identifier is part of the transformer source. At the top of the file mentioned is

```

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var ts = require('typescript');
var ts__default = _interopDefault(ts);
var child_process = require('child_process');
var path = _interopDefault(require('path'));
```
So I'm guessing this has something to do with CJS/ESM compatibility? I'll keep researching along those lines.

As I'm not very familiar with ts transformations, I'm not sure where the problem lies. Can anyone provide some pointers?

If it's a problem in the ts-runtime-checks package, I'd like to submit a PR with a fix, if only I knew what that was! :-)

Looking forward to expanding my knowledge here. Thanks!

**Comment by m-mcgowan** [comment]
The value of `ts__default` is

```
ts__default {
  ANONYMOUS: [Getter],
  AccessFlags: [Getter],
  AssertionLevel: [Getter],
  AssignmentDeclarationKind: [Getter],
  AssignmentKind: [Getter],
  Associativity: [Getter],
  BreakpointResolver: [Getter],
  BuilderFileEmit: [Getter],
  BuilderProgramKind: [Getter],
  BuilderState: [Getter],
  BundleFileSectionKind: [Getter],
  CallHierarchy: [Getter],
  CharacterCodes: [Getter],
  CheckFlags: [Getter],
  CheckMode: [Getter],
  ClassificationType: [Getter],
  ClassificationTypeNames: [Getter],
  CommentDirectiveType: [Getter],
  Comparison: [Getter],
  CompletionInfoFlags: [Getter],
  CompletionTriggerKind: [Getter],
  Completions: [Getter],
  ContainerFlags: [Getter],
  ContextFlags: [Getter],
  Debug: [Getter],
  DiagnosticCategory: [Getter],
  Diagnostics: [Getter],
  DocumentHighlights: [Getter],
  ElementFlags: [Getter],
  EmitFlags: [Getter],
  EmitHint: [Getter],
  EmitOnly: [Getter],
  EndOfLineState: [Getter],
  EnumKind: [Getter],
  ExitStatus: [Getter],
  ExportKind: [Getter],
  Extension: [Getter],
  ExternalEmitHelpers: [Getter],
  FileIncludeKind: [Getter],
  FilePreprocessingDiagnosticsKind: [Getter],
  FileSystemEntryKind: [Getter],
  FileWatcherEventKind: [Getter],
  FindAllReferences: [Getter],
  FlattenLevel: [Getter],
  FlowFlags: [Getter],
  ForegroundColorEscapeSequences: [Getter],
  FunctionFlags: [Getter],
  GeneratedIdentifierFlags: [Getter],
  GetLiteralTextFlags: [Getter],
  GoToDefinition: [Getter],
  HighlightSpanKind: [Getter],
  IdentifierNameMap: [Getter],
  IdentifierNameMultiMap: [Getter],
  ImportKind: [Getter],
  ImportsNotUsedAsValues: [Getter],
  IndentStyle: [Getter],
  IndexFlags: [Getter],
  IndexKind: [Getter],
  InferenceFlags: [Getter],
  InferencePriority: [Getter],
  InlayHintKind: [Getter],
  InlayHints: [Getter],
  InternalEmitFlags: [Getter],
  InternalSymbolName: [Getter],
  InvalidatedProjectKind: [Getter],
  JSDocParsingMode: [Getter],
  JsDoc: [Getter],
  JsTyping: [Getter],
  JsxEmit: [Getter],
  JsxFlags: [Getter],
  JsxReferenceKind: [Getter],
  LanguageServiceMode: [Getter],
  LanguageVariant: [Getter],
  LexicalEnvironmentFlags: [Getter],
  ListFormat: [Getter],
  LogLevel: [Getter],
  MemberOverrideStatus: [Getter],
  ModifierFlags: [Getter],
  ModuleDetectionKind: [Getter],
  ModuleInstanceState: [Getter],
  ModuleKind: [Getter],
  ModuleResolutionKind: [Getter],
  ModuleSpecifierEnding: [Getter],
  NavigateTo: [Getter],
  NavigationBar: [Getter],
  NewLineKind: [Getter],
  NodeBuilderFlags: [Getter],
  NodeCheckFlags: [Getter],
  NodeFactoryFlags: [Getter],
  NodeFlags: [Getter],
  NodeResolutionFeatures: [Getter],
  ObjectFlags: [Getter],
  OperationCanceledException: [Getter],
  OperatorPrecedence: [Getter],
  OrganizeImports: [Getter],
  OrganizeImportsMode: [Getter],
  OuterExpressionKinds: [Getter],
  OutliningElementsCollector: [Getter],
  OutliningSpanKind: [Getter],
  OutputFileType: [Getter],
  PackageJsonAutoImportPreference: [Getter],
  PackageJsonDependencyGroup: [Getter],
  PatternMatchKind: [Getter],
  PollingInterval: [Getter],
  PollingWatchKind: [Getter],
  PragmaKindFlags: [Getter],
  PrivateIdentifierKind: [Getter],
  ProcessLevel: [Getter],
  ProgramUpdateLevel: [Getter],
  QuotePreference: [Getter],
  RelationComparisonResult: [Getter],
  Rename: [Getter],
  ScriptElementKind: [Getter],
  ScriptElementKindModifier: [Getter],
  ScriptKind: [Getter],
  ScriptSnapshot: [Getter],
  ScriptTarget: [Getter],
  SemanticClassificationFormat: [Getter],
  SemanticMeaning: [Getter],
  SemicolonPreference: [Getter],
  SignatureCheckMode: [Getter],
  SignatureFlags: [Getter],

...
}

And sure enough, `createIdentifier` isn't there, although there's a whole host of transformation-related symbols.

**Comment by m-mcgowan** [comment]
`ts.createIdentifier` was deprecated, and eventually removed in 5.0   

```
This release removes many long-deprecated parts of our public API, including (but not limited to):
The top-level Node factory functions (deprecated since TS 4.0) such as ts.createIdentifier; use the factory provided in your TransformationContext or ts.factory instead.
```

From https://github.com/microsoft/TypeScript/wiki/API-Breaking-Changes

**Comment by m-mcgowan** [comment]
I'll close this issue, since it's now clear to me that the problem is the transformer using removed APIs.

**Comment by samchon** [comment]
@m-mcgowan Recommend alternative library `typia`.

https://github.com/samchon/typia

**Comment by m-mcgowan** [comment]
> @m-mcgowan Recommend alternative library `typia`.
> 
> https://github.com/samchon/typia

Thank you for the pointer, I took a brief look and it seems very promising!

**Comment by nonara** [comment]
Hey Matthew!

Fortunately, that's actually an easy fix.

Formerly node creation factory functions were in the larger `ts` namespace, but they were moved to be housed within the `factory` object that is provided by the transformation context.

So youd get the factory and swap to `factory.createIdentifier`. The same will be true for other create/update actions.

Here's an example:

https://github.com/nonara/ts-patch/blob/master/projects/patch/plugin.ts#L21-L41

**Comment by m-mcgowan** [comment]
Thank you for the tip. Once I dies into the code I did try to change things to use and you factory functions but there was also a layer of abstraction and assumptions in the library. It was simpler just to use typia. I managed to implement that is just a few minutes and I have to say I'm thoroughly impressed with all the author has done. 

Thank you again for your support and if I can contribute to typescript or ts-patch at some point I won't hesitate to do that. All the best!
